File: web/src/domain/__tests__/OutputLatency.test.ts
import { describe, expect, it } from 'vitest'
import { HighlightClock } from '../HighlightClock'
import { HighlightEngine } from '../HighlightEngine'
import {
  A2DP_MS,
  LE_MS,
  LOCAL_MS,
  OutputKind,
  OutputLatency,
  OutputRoute,
  classify,
  heardMs,
  highlightMs,
  latencyMs,
  latencyMsForRoute,
} from '../OutputLatency'
import type { Segment } from '../../data/models'

describe('OutputLatency', () => {
  it('empty kinds are local with zero latency', () => {
    expect(classify(new Set())).toBe(OutputRoute.LOCAL)
    expect(latencyMs(new Set())).toBe(0)
  })

  it('local only is zero latency', () => {
    expect(classify(new Set([OutputKind.LOCAL]))).toBe(OutputRoute.LOCAL)
    expect(latencyMsForRoute(OutputRoute.LOCAL)).toBe(LOCAL_MS)
  })

  it('A2DP wins over speaker still listed as an output', () => {
    const kinds = new Set([OutputKind.LOCAL, OutputKind.BLUETOOTH_A2DP])
    expect(classify(kinds)).toBe(OutputRoute.BLUETOOTH_A2DP)
    expect(latencyMs(kinds)).toBe(A2DP_MS)
  })

  it('LE is used when no classic A2DP is present', () => {
    const kinds = new Set([OutputKind.LOCAL, OutputKind.BLUETOOTH_LE])
    expect(classify(kinds)).toBe(OutputRoute.BLUETOOTH_LE)
    expect(latencyMs(kinds)).toBe(LE_MS)
  })

  it('A2DP wins over LE when both are present', () => {
    const kinds = new Set([
      OutputKind.BLUETOOTH_A2DP,
      OutputKind.BLUETOOTH_LE,
      OutputKind.LOCAL,
    ])
    expect(classify(kinds)).toBe(OutputRoute.BLUETOOTH_A2DP)
    expect(latencyMs(kinds)).toBe(A2DP_MS)
  })

  it('heardMs subtracts latency and never goes negative', () => {
    expect(heardMs(1_000, 180)).toBe(820)
    expect(heardMs(50, 180)).toBe(0)
    expect(heardMs(1_000, 0)).toBe(1_000)
  })

  it('highlightMs advances the clock by lead after lag', () => {
    expect(highlightMs(3_800, 0, 1_200)).toBe(5_000)
    expect(highlightMs(1_000, 180, 1_200)).toBe(2_020)
    expect(highlightMs(1_000, 180, 0)).toBe(heardMs(1_000, 180))
    expect(highlightMs(50, 180, 0)).toBe(0)
  })

  it('word lead does not move the heard clock used by non-word surfaces', () => {
    expect(heardMs(1_000, 180)).toBe(820)
    expect(highlightMs(1_000, 180, 1_200)).toBe(2_020)
  })

  it('word lead cannot cross encoded opening silence', () => {
    const segments: Segment[] = [
      { position: 1, startMs: 1_179, endMs: 2_094 },
      { position: 2, startMs: 2_094, endMs: 2_814 },
    ]
    const duringSilence = highlightMs(1_100, 0, 114, segments[0]!.startMs)
    expect(duringSilence).toBe(1_100)
    expect(HighlightEngine.activeWord(segments, duringSilence)).toBeNull()

    const voiceStart = highlightMs(1_179, 0, 114, segments[0]!.startMs)
    expect(voiceStart).toBe(1_179)
    expect(HighlightEngine.activeWord(segments, voiceStart)).toBe(1)
  })

  it('word lead ramps in after the first word so handoff settle cannot skip it', () => {
    const gate = 210
    const lead = 114
    expect(highlightMs(210, 0, lead, gate)).toBe(210)
    expect(highlightMs(260, 0, lead, gate)).toBe(310)
    expect(highlightMs(324, 0, lead, gate)).toBe(438)
    expect(highlightMs(260, 0, lead, 0)).toBe(374)
  })

  it('ramped lead plus settle lights word 1 after media-item handoff', () => {
    const segments: Segment[] = [
      { position: 1, startMs: 210, endMs: 490 },
      { position: 2, startMs: 490, endMs: 1_340 },
    ]
    const clock = new HighlightClock()
    const words: Array<number | null> = []
    for (let tick = 0; tick < 25; tick++) {
      const pos = tick * 33
      const raw = highlightMs(pos, 0, 114, segments[0]!.startMs)
      words.push(HighlightEngine.activeWord(segments, clock.sample('ayah7', raw)))
    }
    expect(words.some((w) => w === 1)).toBe(true)
    expect(words.find((w) => w != null)).toBe(1)
  })

  it('preset values are the documented table', () => {
    expect(OutputLatency.LOCAL_MS).toBe(0)
    expect(OutputLatency.A2DP_MS).toBe(180)
    expect(OutputLatency.LE_MS).toBe(80)
  })
})

File: app/src/test/java/com/beautifulquran/domain/OutputLatencyTest.kt
package com.beautifulquran.domain

import com.beautifulquran.data.model.Segment
import com.beautifulquran.domain.OutputLatency.OutputKind
import com.beautifulquran.domain.OutputLatency.Route
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test

class OutputLatencyTest {

    @Test
    fun `empty kinds are local with zero latency`() {
        assertEquals(Route.LOCAL, OutputLatency.classify(emptySet()))
        assertEquals(0L, OutputLatency.latencyMs(emptySet()))
    }

    @Test
    fun `local only is zero latency`() {
        assertEquals(Route.LOCAL, OutputLatency.classify(setOf(OutputKind.LOCAL)))
        assertEquals(OutputLatency.LOCAL_MS, OutputLatency.latencyMs(Route.LOCAL))
    }

    @Test
    fun `A2DP wins over speaker still listed as an output`() {
        val kinds = setOf(OutputKind.LOCAL, OutputKind.BLUETOOTH_A2DP)
        assertEquals(Route.BLUETOOTH_A2DP, OutputLatency.classify(kinds))
        assertEquals(OutputLatency.A2DP_MS, OutputLatency.latencyMs(kinds))
    }

    @Test
    fun `LE is used when no classic A2DP is present`() {
        val kinds = setOf(OutputKind.LOCAL, OutputKind.BLUETOOTH_LE)
        assertEquals(Route.BLUETOOTH_LE, OutputLatency.classify(kinds))
        assertEquals(OutputLatency.LE_MS, OutputLatency.latencyMs(kinds))
    }

    @Test
    fun `A2DP wins over LE when both are present`() {
        val kinds = setOf(
            OutputKind.BLUETOOTH_A2DP,
            OutputKind.BLUETOOTH_LE,
            OutputKind.LOCAL,
        )
        assertEquals(Route.BLUETOOTH_A2DP, OutputLatency.classify(kinds))
        assertEquals(OutputLatency.A2DP_MS, OutputLatency.latencyMs(kinds))
    }

    @Test
    fun `heardMs subtracts latency and never goes negative`() {
        assertEquals(820L, OutputLatency.heardMs(1_000L, 180L))
        assertEquals(0L, OutputLatency.heardMs(50L, 180L))
        assertEquals(1_000L, OutputLatency.heardMs(1_000L, 0L))
    }

    @Test
    fun `highlightMs advances the clock by lead after lag`() {
        // Word timed at 5000 should light when media is at 3800 with 1200 lead.
        assertEquals(5_000L, OutputLatency.highlightMs(3_800L, latencyMs = 0L, leadMs = 1_200L))
        // Lag and lead net: 1000 − 180 + 1200 = 2020.
        assertEquals(2_020L, OutputLatency.highlightMs(1_000L, latencyMs = 180L, leadMs = 1_200L))
        // Zero lead matches heardMs.
        assertEquals(
            OutputLatency.heardMs(1_000L, 180L),
            OutputLatency.highlightMs(1_000L, latencyMs = 180L, leadMs = 0L),
        )
        assertEquals(0L, OutputLatency.highlightMs(50L, latencyMs = 180L, leadMs = 0L))
    }

    @Test
    fun `word lead does not move the heard clock used by non-word surfaces`() {
        assertEquals(820L, OutputLatency.heardMs(1_000L, latencyMs = 180L))
        assertEquals(
            2_020L,
            OutputLatency.highlightMs(1_000L, latencyMs = 180L, leadMs = 1_200L),
        )
    }

    @Test
    fun `word lead cannot cross encoded opening silence`() {
        val segments = listOf(
            Segment(position = 1, startMs = 1_179, endMs = 2_094),
            Segment(position = 2, startMs = 2_094, endMs = 2_814),
        )
        val duringSilence = OutputLatency.highlightMs(
            mediaPositionMs = 1_100,
            latencyMs = 0,
            leadMs = 114,
            leadNotBeforeMs = segments.first().startMs,
        )
        assertEquals(1_100L, duringSilence)
        assertEquals(null, HighlightEngine.activeWord(segments, duringSilence))

        val voiceStart = OutputLatency.highlightMs(
            mediaPositionMs = 1_179,
            latencyMs = 0,
            leadMs = 114,
            leadNotBeforeMs = segments.first().startMs,
        )
        // At the gate, ramped lead is still 0 — continuous with silence.
        assertEquals(1_179L, voiceStart)
        assertEquals(1, HighlightEngine.activeWord(segments, voiceStart))
    }

    @Test
    fun `word lead ramps in after the first word so handoff settle cannot skip it`() {
        // Hard +lead at the gate (210 → 324) exceeds HighlightClock.MAX_SETTLE_STEP_MS
        // (100) and freezes the post-ayah-handoff clock through a short word 1.
        val gate = 210L
        val lead = 114L
        assertEquals(
            210L,
            OutputLatency.highlightMs(210, latencyMs = 0, leadMs = lead, leadNotBeforeMs = gate),
        )
        // Half-way through the ramp: applied lead == pastGate.
        assertEquals(
            310L, // 260 + min(114, 50)
            OutputLatency.highlightMs(260, latencyMs = 0, leadMs = lead, leadNotBeforeMs = gate),
        )
        // Full lead once pastGate >= leadMs.
        assertEquals(
            438L, // 324 + 114
            OutputLatency.highlightMs(324, latencyMs = 0, leadMs = lead, leadNotBeforeMs = gate),
        )
        // No gate: full lead from the first sample (unchanged).
        assertEquals(
            374L,
            OutputLatency.highlightMs(260, latencyMs = 0, leadMs = lead, leadNotBeforeMs = 0),
        )
    }

    @Test
    fun `ramped lead plus settle lights word 1 after media-item handoff`() {
        // Hani 3:7-shaped row: short word 1 inside the post-handoff settle window.
        val segments = listOf(
            Segment(position = 1, startMs = 210, endMs = 490),
            Segment(position = 2, startMs = 490, endMs = 1_340),
        )
        val clock = HighlightClock()
        val words = mutableListOf<Int?>()
        for (tick in 0 until 25) {
            val pos = tick * 33L
            val raw = OutputLatency.highlightMs(
                mediaPositionMs = pos,
                latencyMs = 0,
                leadMs = 114,
                leadNotBeforeMs = segments.first().startMs,
            )
            val key = "ayah7"
            val ms = clock.sample(key, raw)
            words.add(HighlightEngine.activeWord(segments, ms))
        }
        assertTrue(
            "word 1 must light before word 2 after handoff; saw $words",
            words.any { it == 1 },
        )
        val firstLit = words.first { it != null }
        assertEquals(1, firstLit)
    }

    @Test
    fun `preset values are the documented table`() {
        assertEquals(0L, OutputLatency.LOCAL_MS)
        assertEquals(180L, OutputLatency.A2DP_MS)
        assertEquals(80L, OutputLatency.LE_MS)
    }
}
